### PR TITLE
moshi handling for lists and maps of wire messages

### PR DESCRIPTION
**Pull request by mmihic**
Previously we were relying on moshi's default handling for repeated and
map fields on wire messages; this is fine for primitives, but populates
the collection with the wrong types when dealing with repeated or maps
of other wire messages. This adds proper support for dealing with those
types of collections.

### CODE REVIEW

File: misk/src/main/kotlin/misk/moshi/wire/ListAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    override fun fromJson(reader: JsonReader): List<Any?>? {
+        val nextToken = reader.peek()
+        if (nextToken == JsonReader.Token.NULL) return reader.nextNull()
+        check(nextToken == JsonReader.Token.BEGIN_ARRAY) {
+            "expected [; found ${reader.peek()}"
+        }
+
+        val elements = mutableListOf<Any?>()
+        reader.beginArray()
+        while (reader.peek() != JsonReader.Token.END_ARRAY) {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/moshi/wire/ListAdapter.kt]: reader.hasNext()

File: misk/src/main/kotlin/misk/moshi/wire/ListAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import com.squareup.moshi.JsonAdapter
+import com.squareup.moshi.JsonReader
+import com.squareup.moshi.JsonWriter
+
+/** Handles JSON marshalling / unmarshalling for type-safe lists */
+internal class ListAdapter(private val elementAdapter: JsonAdapter<Any>) :
+        JsonAdapter<List<Any?>>() {
+    override fun fromJson(reader: JsonReader): List<Any?>? {
+        val nextToken = reader.peek()
+        if (nextToken == JsonReader.Token.NULL) return reader.nextNull()
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/moshi/wire/ListAdapter.kt]: If you want, you get this for free by calling .nullSafe on this adapter

File: misk/src/main/kotlin/misk/moshi/wire/MapAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        if (nextToken == JsonReader.Token.NULL) return reader.nextNull()
+        check(nextToken == JsonReader.Token.BEGIN_OBJECT) {
+            "expected {, found $nextToken"
+        }
+
+        val results = mutableMapOf<Any, Any>()
+        reader.beginObject()
+        while (reader.peek() != JsonReader.Token.END_OBJECT) {
+            val keyName = reader.nextName()
+            val key = keyConverter(keyName)
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/moshi/wire/MapAdapter.kt]: Nice.

Comment (swankjesse) [inline misk/src/main/kotlin/misk/moshi/wire/MapAdapter.kt]: Off-topic: this syntax makes perfect sense but it's still surprising. I'm still learning apparently!

File: misk/src/test/kotlin/misk/moshi/wire/WireMessageAdapterTest.kt [hunk 4]
```diff
@@ -136,6 +168,43 @@ internal class WireMessageAdapterTest {
...
+                .build()
+        val json = warehouseAdapter.indent(" ").toJson(warehouse)
+        assertThat(json).isEqualToAsJson("""
+{
+  "warehouse_id": 1976,
+  "warehouse_token": "W_ACDFD",
+  "robots": {
+    "34": {
+      "robot_id": 34,
+      "robot_token": "R_93498"
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/moshi/wire/WireMessageAdapterTest.kt]: Oh Good old R_93498. He was a good robot.

**Comment by ryanhall07** [comment]
Is this something Moshi should support? If so, should we make that clear with an explicit package so it's not jumbled with other Misk-specific Moshi code?

**Comment by mmihic** [comment]
It's specific to moshi handling of wire (like the other classes in this package)

**Comment by ryanhall07** [comment]
Although Moshi is Java, so it wouldn't be a direct contribution to Moshi :/

**Pull request merged**